Add OPNsense to Router (#1168)
pi-hole/docs: docs/routers/OPNsense.md

> This guide was developed using a OPNsense 25.1.1. Following the first method below will have you adding your Pi-hole as a DNS server for all devices on your LAN. Following the second method below will have DNS queries route through your USG and then to your Pi-hole. Either option is valid, depending on your specific requirements, but it is recommended to choose one or the other - not both.

Reviewer: > route through your USG

This seems to be a leftover

Author: Thank you for taking a look at it. I hope I did the squash correctly (first time using it).

Author: Is it ok for me to resolve this conversation?
Or is it good practice that you do that instead?

There is also the different firewall rule options to prevent clients like Chromecasts of using their hardcoded 8.8.8.8 DNS servers and use your pihole Pi-hole instead.
Beware: This tutorial does not explain on how to block DoH or DoT.

### Grab your IPv4 and IPv6 address from your Pi-hole

SSH into your Pi-hole. A quick `ifconfig eth0` (substituting `eth0` for the applicable network interface) will get you the address information:

```
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.0.51.2  netmask 255.255.255.0  broadcast 192.168.0.255
        inet6 2001:470:5:3::8ed  prefixlen 128  scopeid 0x0<global>
        inet6 2001:470:5:3:8d70:e92e:7030:6eef  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::337f:710d:6781:4c85  prefixlen 64  scopeid 0x20<link>
        ether dc:a6:32:0e:31:2d  txqueuelen 1000  (Ethernet)
        RX packets 1297459  bytes 1279521260 (1.1 GiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 2315988  bytes 2766370095 (2.5 GiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

> Take note of the `inet` address, and the `inet6` address which starts with `fe` (denotes a local link or LAN address). Those are the IPs which you'll use as your DNS servers in the USG. For this example, we'll be using the `10.0.51.2` and `fe80::337f:710d:6781:4c85` IPs for the following directions.

Reviewer: USG mentionted here

### Method 1: Distribute Pi-hole as DNS server via DHCP

Choosing this option will have DNS queries taking the following path:

```

> (Clients) -> Pi-hole -> Upstream DNS Server (or Unbound reverse DNS)

Reviewer: Why did you include Unbound here? Your Guide does not reference to Unbound otherwise.

Author: Well simply because it is a possibility. Do you think it distracts more than it helps and it should be removed?

Reviewer: Yes, I think it distracts a bit. It looks like, Unbound would not be an upstream server, but something else. Btw. it's not a reverse DNS server but a recursive :-)

Author: I removed it

<img width="1576" alt="image" src="https://github.com/user-attachments/assets/670b8041-a889-4335-bbb8-445743141770" />

#### IPv6

Go to **Interfaces** and click on the interface that you want to modify.
Under **Manual configuration** check "Allow manual adjustment of DHCPv6 and Router Advertisements"

<img width="1025" alt="image" src="https://github.com/user-attachments/assets/65f73a6f-ee37-49d1-b3a0-8fbee13a6fa6" />

Now go to **Services** --> **Router Advertisement** and click on the interface that you want to modify.
Under **DNS servers** enter your Raspi's IPv6 address into the first line.
Leave the second line blank unless you have another Pi-hole host running.

<img width="957" alt="image" src="https://github.com/user-attachments/assets/d427a5d0-9bd7-4e31-b7e6-7e8251706e5c" />

Beware: Link local IPv6 are not reachable from other VLANs.

All client devices on your network should now automatically be configured with a single IPv4 and single IPv6 address for DNS resolution. If you don't see your client devices update quickly enough, you *can* toggle Wi-Fi off/on or disconnect the Ethernet cable for a couple of seconds and then reconnect it.

> ### Method 2: Using your Pi-hole as DNS server for your OPNsense

Reviewer: Maybe a Screenshot from the interface would be nice for this section?

Author: I think that is pretty hard, since there are so many possible clients and because of that interfaces. 
macOS, Linux, iOS, Android, Windows...

Author: ahh got it. Added a picture

> If you'd rather have your OPNsense forward all DNS traffic to Pi-hole, you can complete this *alternative* step. Unless manually configured otherwise, all devices on your network will use Pi-hole as their DNS server.

Reviewer: > nless manually configured otherwise, all devices on your network will use Pi-hole as their DNS server.

This sentence doesn't make sense when you the sentence rigth before says all Clients will user OPNSense als their upstream DNS server

Author: Yeah, I copied it from the USG one. 

It does make sense, but it is easy to misunderstand. 
If you set OPNsense to use pihole as DNS, clinets will ask OPNsense, which then will ask pi-hole. 
So in the and all devices will use Pihole (even if indirect) UNLESS the have for example the setting 8.8.8.8.

But I think it distracts more than it helps and will delete it.

Reviewer: > Yeah, I copied it from the USG one.

Correct, but the second sentence is new to this guide.

Doing so will have DNS queries taking the following path:

```
(Clients) -> OPNsense -> Pi-hole -> Upstream DNS Server
```

!!! warning
    Keep in mind that if you are using your OPNsense IP address as an upstream resolver for your Pi-hole together with this configuration, ***this will cause a resolver loop***. Make sure to have another upstream DNS server(s) set in your Pi-hole's configuration, or use Pi-hole as [your own recursive DNS server with Unbound](https://docs.pi-hole.net/guides/dns/unbound/?h=unbound#configure-unbound).

1. In Pi-hole, navigate to `Settings -> DNS` and ensure you have at least one external upstream DNS server enabled.
2. In OPNsense navigate to `Settings -> General -> Networking`.
3. Under DNS Servers, enter the IPv4 and or IPv6 address of your Pi-hole host, and set the gateway to your WAN interface.
4. Uncheck `Allow DNS server list to the overridden by DHCP/PPP on WAN`.
5. Click Save.

> ### Block other DNS servers (optional)

Reviewer: Please add a warning, that this might break some devices functionality if they require hard-coded DNS servers to respond.

Author: Do you have an example?

Only device I have ever encountered that uses hardcoded DNS was Chromecast. 
And there it was not an issue, since they use DHCP DNS as fallback. 

I would argue that if you hardcode a DNS setting, you are also responsible for a fallback.

Reviewer: There are certain Android Apps, Google devices like Chromecast, Ring, Panasonic TVs,..

https://discourse.pi-hole.net/search?q=hardcoded%20dns

> I would argue that if you hardcode a DNS setting, you are also responsible for a fallback.

I agree, but we want to be kind and and help users where other developers might 'forget' a fallback

This guide was developed using a OPNsense 25.1.1. Following the first method below will have you adding your Pi-hole as a DNS server for all devices on your LAN. Following the second method below will have DNS queries route through your OPNsense and then to your Pi-hole. Either option is valid, depending on your specific requirements, but it is recommended to choose one or the other - not both.

> There is also the different firewall rule options to prevent clients like Chromecasts of using their hardcoded 8.8.8.8 DNS servers and use your pihole Pi-hole instead.

Reviewer: ```suggestion
Optionally, there is also the possibility to block DNS traffic to other, non-Pi-hole DNS servers via firewall rules. This can be used to prevent clients like Chromecasts of using their hardcoded DNS servers.
```

I would not go so far to safely assume they will use Pi-hole instead. As we discussed, this might just break DNS for them. If you really want to force them to use Pi-hole, you need to enforce a routing rule to redirect all DNS traffic to Pi-hole.

### Method 1: Distribute Pi-hole as DNS server via DHCP

Choosing this option will have DNS queries taking the following path:

```
(Clients) -> Pi-hole -> Upstream DNS Server (or Unbound reverse DNS)
```

#### IPv4

Go to **Services** --> **ISC DHCPv4** and click on the interface that you want to modify.
Under **DNS servers** enter your Pi-hole IPv4 address into the first line.
Leave the second line blank unless you have another Pi-hole host running.

![Screenshot of OPNsense DHCPv4 Settings](../images/routers/opnsense-dhcpv4.png)

#### IPv6

Go to **Interfaces** and click on the interface that you want to modify.
Under **Manual configuration** check "Allow manual adjustment of DHCPv6 and Router Advertisements"

![Screenshot of OPNsense IPv6 manual RA Settings](../images/routers/opnsense-ipv6-manual-ra.png)

Now go to **Services** --> **Router Advertisement** and click on the interface that you want to modify.
Under **DNS servers** enter your Pi-Holes IPv6 address into the first line.
Leave the second line blank unless you have another Pi-hole host running.

> ![Screenshot of OPNsense IPv6 RA Settings](../images/routers/opnsense-ipv6-manual-ra.png)

Reviewer: I think you wanted to include a different image here

Author: corrected it

Doing so will have DNS queries taking the following path:

```
(Clients) -> OPNsense -> Pi-hole -> Upstream DNS Server
```

!!! warning
    Keep in mind that if you are using your OPNsense IP address as an upstream resolver for your Pi-hole together with this configuration, ***this will cause a resolver loop***. Make sure to have another upstream DNS server(s) set in your Pi-hole's configuration, or use Pi-hole as [your own recursive DNS server with Unbound](https://docs.pi-hole.net/guides/dns/unbound/?h=unbound#configure-unbound).

1. In Pi-hole, navigate to `Settings -> DNS` and ensure you have at least one external upstream DNS server enabled.
2. In OPNsense navigate to `Settings -> General -> Networking`.
3. Under DNS Servers, enter the IPv4 and or IPv6 address of your Pi-hole host, and set the gateway to your WAN interface.
4. Uncheck `Allow DNS server list to the overridden by DHCP/PPP on WAN`.
5. Click Save.

### Block other DNS servers (optional)

Some clients have hardcoded DNS settings. To prevent them from circumventing our Pi-Hole, we can optionally block all DNS request that are not coming from our Pi-Hole.

> Warning, if your device does not have any fallback from the hardcoded DNS settings, this could break it from resolving.

Reviewer: ```suggestion
!!! warning
   If your device does not have any fallback from the hardcoded DNS settings, this could break its DNS.
```

We first create an Alias for Pi-holes IPs.
Go to **Firewall** --> **Aliases** scroll down and click on the plus button to create a new alias.
Give it a Name and add the IPv4 and IPv6 into the "Content" field and then click save.

![Screenshot of OPNsense Aliases](../images/routers/opnsene-alias.png)

Now go to **Firewall** --> **Rules** and click on the interface that you want to modify.

> We create one rule to allow pihole to do DNS over IPv4 + IPv6.

Reviewer: ```suggestion
We create one rule to allow Pi-hole to do DNS over IPv4 + IPv6.
```

After that rule, we create a rule to reject ALL DNS. It is important that this reject rule is after the other rule.
OPNsense does the first rule that matches, if we have our reject rule first, Pi-hole can not reach any DNS servers.
It should look something like this:

![Screenshot of OPNsense Firewall rules](../images/routers/opnsense-firewall-rule.png)

> We don't need a rule to allow for example device 10.0.51.10 to reach our pihole at 10.0.51.2, since they are both on the same network, DNS request will not be routed through OPNsense. Instead the client 10.0.51.10 will directly contact Pi-hole for DNS requests.

Reviewer: ```suggestion
We don't need a rule to allow for example device 10.0.51.10 to reach our Pi-hole at 10.0.51.2, since they are both on the same network, DNS request will not be routed through OPNsense. Instead the client 10.0.51.10 will directly contact Pi-hole for DNS requests.
```
